File: packages/json/llm-json-stream-ts/packages/llm-json-stream/src/classes/property_delegates/null_property_delegate.ts
/**
 * Delegate for parsing JSON null values.
 */

import { PropertyDelegate } from "./property_delegate.js";
import { JsonStreamParserController } from "../json_stream_parser.js";

export class NullPropertyDelegate extends PropertyDelegate {
    private buffer = "";

    constructor(
        propertyPath: string,
        parserController: JsonStreamParserController,
        onComplete?: () => void,
    ) {
        super(propertyPath, parserController, onComplete);
    }

    addCharacter(character: string): void {
        // Check for delimiters that end the null
        if (character === "," || character === "}" || character === "]") {
            if (!this.isDone && this.buffer === "null") {
                this.isDone = true;
                this.parserController.completeProperty(this.propertyPath, null);
                this.onComplete?.();
            }
            return;
        }

        // Build the buffer
        this.buffer += character;

        // Check if we've completed parsing null
        if (this.buffer === "null") {
            this.isDone = true;
            this.parserController.completeProperty(this.propertyPath, null);
            this.onComplete?.();
        }
    }
}

File: packages/json/llm-json-stream-python/src/llm_json_stream/property_delegates/null_property_delegate.py
"""Delegate for parsing JSON null values."""

from __future__ import annotations

from ..property_delegate import PropertyDelegate


class NullPropertyDelegate(PropertyDelegate):
    def __init__(self, property_path: str, parser_controller, on_complete=None) -> None:
        super().__init__(property_path, parser_controller, on_complete)
        self._buffer = ""

    def add_character(self, character: str) -> None:
        if character in {",", "}", "]"}:
            if not self._is_done and self._buffer == "null":
                self._is_done = True
                self._parser_controller.complete_property(self._property_path, None)
                if self._on_complete:
                    self._on_complete()
            return

        self._buffer += character
        if self._buffer == "null":
            self._is_done = True
            self._parser_controller.complete_property(self._property_path, None)
            if self._on_complete:
                self._on_complete()
